File: ViewAutomata/statofa.cpp
#include "statofa.h"

StatoFA::StatoFA(const std::string & nome, const bool& finale):Stato(nome,finale)
{
    trans = new std::vector<Transizione*>();
}

Transizione *StatoFA::getTrans(StatoFA * stato, const char & input)
{
    std::vector<Transizione*>::iterator i;
    for(i = trans->begin(); i != trans->end() && ((*i)->getDest() != stato || (*i)->getInput() != input); i++);

    if(i == trans->end())
        return nullptr;
    else
        return (*i);
}

void StatoFA::add(StatoFA * stato, const char & input)
{
    if(getTrans(stato,input) == nullptr)
    {
        trans->push_back(new Transizione(stato,input));
    }
}

void StatoFA::remove(StatoFA * stato, const char & input)
{
    bool eliminato = false;
    for(std::vector<Transizione*>::iterator j = trans->begin();  j != trans->end() && !eliminato; j++)
    {
        if((*j)->getDest() == stato && (*j)->getInput() == input)
        {
            trans->erase(j);
            eliminato = true;
        }
    }
}

int StatoFA::nTrans()
{
    return trans->size();
}

Transizione *StatoFA::operator [](int i)
{
    return trans->at(i);
}

File: JAVA/src/progettoP2/StatoFA.java


import java.util.Vector;

public class StatoFA extends Stato{
	private Vector<Transizione> trans = new Vector<Transizione>();
    public StatoFA(final String nome, final Boolean finale){
    	super(nome, finale);
    }
    public StatoFA(String s) {
		super(s,false);
	}
    
	public Transizione getTrans(StatoFA stato, final char input){
    	int j = 0;
        while(j<trans.size() && (trans.get(j).getDest() != stato 
        		|| trans.get(j).getInput() != input))
            j++;
        if(j == trans.size())
            return null;
        else
            return trans.get(j);
    }
    public void add(StatoFA stato, final char input){
    	if(getTrans(stato,input) == null)
        {
            trans.addElement(new Transizione(stato,input));
        }
    }
    public void remove(StatoFA stato, final char input){
    	Boolean eliminato = false;
        for(int j = 0;  j < trans.size() && !eliminato; j++)
        {
            if(trans.get(j).getDest() == stato && trans.get(j).getInput() == input)
            {
                trans.remove(j);
                eliminato = true;
            }
        }
    }
    public int nTrans(){
    	return trans.size();
    }
    public Transizione operatorBarraBarra(int i){
    	return trans.get(i);
    }
}
